core/src/main/java/org/elasticsearch/transport/TcpChannel.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.action.ListenableActionFuture;
+
+import java.io.IOException;
+
+public interface TcpChannel<C extends TcpChannel<C>> {
+
+    ListenableActionFuture<C> closeAsync();
```

Comment:
can we turn this around and accept a listener isntead? I really don't like returning futures it's a design flaw IMO

Reply:
I'm not sure I understand here. We need something to block on at shutdown. Are you suggesting this?

```
PlainActionFuture<C> future = PlainActionFuture.newFuture();
channel.addCloseListener(future);
channel.closeAsync();
future.actionGet();
```

Reply:
Or some callback with a `CountDownLatch`?

```
CountDownLatch latch = new CountDownLatch(channels.size());
        for (final C channel : channels) {
            channel.addCloseListener(ActionListener.wrap(latch::countDown));
        }
        latch.await();
```

Reply:
I think the simplest options would be this:
```Java

void close(ActionListener<Void> onClosed);

```